server/tls/tlscertificate_retriever_file.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 Microsoft. All rights reserved.
+
+package tls
+
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"golang.org/x/crypto/pkcs12"
+	"io/ioutil"
+)
+
+const (
+	CertLabel = "CERTIFICATE"
+	PrivateKeyLabel = "PRIVATE KEY"
+)
+
+type filetlsCertificateRetriever struct {
+	pemBlock []*pem.Block
+	settings TlsSettings
+}
+
+// GetCertificate Returns the certificate associated with the pfx
+func (fcert *filetlsCertificateRetriever) GetCertificate() (*x509.Certificate, error) {
+	for _, block := range fcert.pemBlock {
+		if block.Type == CertLabel {
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, fmt.Errorf("Failed to parse certificate with error %+v", err)
+			}
+			if cert.IsCA != true {
+				return cert, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("No Certificate block found")
+}
+
+// GetPrivateKey Returns the private key associated with the pfx
+func (fcert *filetlsCertificateRetriever) GetPrivateKey() (crypto.PrivateKey, error) {
+
+	for _, block := range fcert.pemBlock {
+		if block.Type == PrivateKeyLabel {
+			pk, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+			if err != nil {
+				return nil, fmt.Errorf("Could not parse private key %+v", err)
+			}
+			return pk, nil
+		}
+	}
+
+	return nil, nil
```

Comment:
when will we return nil, nil ?
1. should we return non-nil error if privateKey is nil ?
2. Can privateKey be nil? is it a pointer?

Reply:
It's actually just a interface{} under the hood, so it can be nil. I have updated this to return a error message. In fact, I thought I did that in a earlier iteration but must of been lost, nice catch.